lib/go/edgecontext/edgecontexttest/runner.go

```diff
@@ -0,0 +1,726 @@
+		t.Errorf(
+			"Expected ExperimentEvent.IsOverride %v, got %v",
+			expected.IsOverride,
+			actual.IsOverride,
+		)
+	}
+
+	if expected.EventType != actual.EventType {
+		t.Errorf(
+			"Expected ExperimentEvent.EventType %v, got %v",
+			expected.EventType,
+			actual.EventType,
+		)
+	}
+}
+
+func compareTouchedFields(
+	t *testing.T,
+	actual experiments.ExperimentEvent,
+	userID string,
+	loggedIn bool,
+	cookieTime time.Time,
+	oauthID string,
+	session string,
+	device string,
+) {
+	t.Helper()
+
+	if userID != actual.UserID {
+		t.Errorf(
+			"Expected ExperimentEvent.UserID %v, got %v",
+			userID,
+			actual.UserID,
+		)
+	}
+
+	if actual.LoggedIn == nil {
+		t.Errorf("Expected non-nil ExperimentEvent.LoggedIn")
+	} else {
+		if loggedIn != *actual.LoggedIn {
+			t.Errorf(
+				"Expected ExperimentEvent.LoggedIn %v, got %v",
+				loggedIn,
+				*actual.LoggedIn,
+			)
+		}
+	}
+
+	if !cookieTime.Equal(actual.CookieCreatedAt) {
+		t.Errorf(
+			"Expected ExperimentEvent.CookieCreatedAt %v, got %v",
+			cookieTime,
+			actual.CookieCreatedAt,
+		)
+	}
+
+	if oauthID != actual.OAuthClientID {
+		t.Errorf(
+			"Expected ExperimentEvent.OAuthClientID %v, got %v",
+			oauthID,
+			actual.OAuthClientID,
+		)
+	}
+
+	if session != actual.SessionID {
+		t.Errorf(
+			"Expected ExperimentEvent.SessionID %v, got %v",
+			session,
+			actual.SessionID,
+		)
+	}
+
+	if device != actual.DeviceID.String() {
+		t.Errorf(
+			"Expected ExperimentEvent.DeviceID %v, got %v",
+			device,
+			actual.DeviceID,
+		)
+	}
+}
+
+func edgeContextWithLocaleCode(l string, impl *edgecontext.Impl) error {
+	ctx := context.Background()
+	_, err := edgecontext.New(
+		ctx,
+		impl,
+		edgecontext.NewArgs{
+			LocaleCode: l,
+		},
+	)
+	return err
+}
+
+type runnerConfig struct {
+	useDefaultBackend bool
+	backend           *edgecontext.Backend
+}
+
+type RunnerOption func(*runnerConfig)
+
+func WithBackend(t *testing.T, backend *edgecontext.Backend) RunnerOption {
+	return func(c *runnerConfig) {
+		c.backend = backend
+	}
+}
+
+func WithDefaultBackend() RunnerOption {
+	return func(c *runnerConfig) {
+		c.useDefaultBackend = true
+	}
+}
+
+type Runner struct {
+	impl *edgecontext.Impl
+}
+
+func NewRunner(t *testing.T, opts ...RunnerOption) *Runner {
+	var cfg runnerConfig
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+	if cfg.backend != nil && !cfg.useDefaultBackend {
+		edgecontext.SetBackend(cfg.backend)
+	} else if cfg.backend == nil && cfg.useDefaultBackend {
+		// nothing to do here, this is just to simplify the configuration
+		// check code and make it clear what cases are valid.
+	} else {
+		t.Fatalf("must specify either backend or use WithDefaultBackend")
+	}
+
+	store, _, err := secrets.NewTestSecrets(
+		context.Background(),
+		map[string]secrets.GenericSecret{
+			"secret/authentication/public-key": {
+				Type:     secrets.VersionedType,
+				Encoding: secrets.IdentityEncoding,
+				Current:  verificationSecret,
+			},
+		},
+	)
+
```

Comment:
style nit: I'd prefer not have a blank line between the code generating `err` and the `if` checking it.